### PR TITLE
[sca] Add SW key masking for KMAC.

### PR DESCRIPTION
**Pull request by vrozic**
Signed-off-by: Vladimir Rozic <vrozic@lowrisc.org>

### CODE REVIEW

File: sw/device/sca/sha3_serial.c [hunk 1]
```diff
@@ -216,14 +256,16 @@ static dif_result_t kmac_msg_start(dif_kmac_mode_kmac_t mode, size_t l,
...
   for (int i = 0; i < ARRAYSIZE(k->share0); ++i) {
+    a = next_lfsr();
     mmio_region_write32(kmac.base_addr,
                         KMAC_KEY_SHARE0_0_REG_OFFSET + i * sizeof(uint32_t),
-                        k->share0[i]);
+                        k->share0[i]^a);
     mmio_region_write32(kmac.base_addr,
                         KMAC_KEY_SHARE1_0_REG_OFFSET + i * sizeof(uint32_t),
-                        k->share1[i]);
+                        k->share1[i]^a);
```
Comment (andreaskurth) [inline sw/device/sca/sha3_serial.c:268]: `a` is generated with the `next_lfsr` function, which is documented to be a PRNG for randomizing the sampling order during batch capture. But I think these lines use `a` to mask the shares of the key, right? Does that work? If so, could you add a comment why this works?

**Review by andreaskurth — COMMENTED** [review]
Thanks for your PR, @vrozic! I have a couple of questions

File: sw/device/sca/sha3_serial.c [hunk 2]
```diff
@@ -155,6 +155,46 @@ static uint32_t calculate_rate_bits(uint32_t security_level) {
...
+}
+
+/**
+ * PRNG for randomizing sampling order during batch capture
+ */
+static uint32_t lfsr_state = 0x5a5a00ff;
+
+uint32_t next_lfsr(void) {
+  const uint32_t lfsr_out = lfsr_state;
+  // If you get too many random keys than this is wrong direction
```
[the remark this answers is not in the document]
Comment (vrozic) [reply sw/device/sca/sha3_serial.c]: Sorry, that's the leftover comment from the testing stage. I wasn't sure about the LFSR shifting direction. It is removed now.

File: sw/device/sca/sha3_serial.c [hunk 3]
```diff
@@ -155,6 +155,46 @@ static uint32_t calculate_rate_bits(uint32_t security_level) {
...
+	s[1] ^= s[2];
+	s[0] ^= s[3];
+
+	s[2] ^= t;
+
+	s[3] = rotl(s[3], 11);
+
+	return result;
+}
+
```
[the remark this answers is not in the document]
Comment (vrozic) [reply sw/device/sca/sha3_serial.c]: That's one of the older PRNGs that I've been testing and comparing.
It is removed in the latest change.

File: sw/device/sca/sha3_serial.c [hunk 4]
```diff
@@ -216,14 +256,16 @@ static dif_result_t kmac_msg_start(dif_kmac_mode_kmac_t mode, size_t l,
   }
 
   // Set key length and shares.
+  uint32_t a;
   mmio_region_write32(kmac.base_addr, KMAC_KEY_LEN_REG_OFFSET, key_len);
   for (int i = 0; i < ARRAYSIZE(k->share0); ++i) {
+    a = next_lfsr();
```
[the remark this answers is not in the document]
Comment (vrozic) [reply sw/device/sca/sha3_serial.c]: Good idea. It's changed now

File: sw/device/sca/lib/sca.c [hunk 5]
```diff
@@ -293,3 +293,17 @@ void sca_call_and_sleep(sca_callee callee, uint32_t sleep_cycles) {
   OT_DISCARD(dif_clkmgr_gateable_clock_set_enabled(
       &clkmgr, CLKMGR_CLK_ENABLES_CLK_IO_DIV4_PERI_EN_BIT, kDifToggleEnabled));
 }
+
+/**
+ * PRNG for for masking key.
```
Comment (andreaskurth) [inline sw/device/sca/lib/sca.c:298]:
```suggestion
 * PRNG for masking key.
```

File: sw/device/sca/lib/sca.h [hunk 6]
```diff
@@ -159,6 +159,11 @@ typedef void (*sca_callee)(void);
  */
 void sca_call_and_sleep(sca_callee callee, uint32_t sleep_cycles);
 
+/**
+ * PRNG for for masking key.
```
Comment (andreaskurth) [inline sw/device/sca/lib/sca.h:163]:
```suggestion
 * PRNG for masking key.
```

**Review by andreaskurth — APPROVED** [review]
Minor typo, otherwise LGTM :+1:

File: sw/device/sca/lib/sca.c (hunk 5 above)
Comment (vrozic) [reply sw/device/sca/lib/sca.c]: Oops, thanks for noticing. It's fixed now.

File: sw/device/sca/lib/sca.h (hunk 6 above)
Comment (vrozic) [reply sw/device/sca/lib/sca.h]: fixed

File: sw/device/sca/lib/sca.c [hunk 7]
```diff
@@ -293,3 +293,17 @@ void sca_call_and_sleep(sca_callee callee, uint32_t sleep_cycles) {
...
+/**
+ * PRNG for masking key.
+ */
+static uint32_t lfsr_state = 0xdeadbeef;
+
+uint32_t next_lfsr(void) {
+  const uint32_t lfsr_out = lfsr_state;
+  bool lfsr_bit = lfsr_state & 0x00000001;
+  lfsr_state = lfsr_state >> 1;
+  if (lfsr_bit)
```
Comment (engdoreis) [inline sw/device/sca/lib/sca.c:306]: Could you add braces to this `if` for complience with the coding [style](https://docs.opentitan.org/doc/rm/c_cpp_coding_style/)?

**Review by engdoreis — APPROVED** [review]
LGTM.

**Review by vogelpi — APPROVED** [review]
LGTM, thanks @vrozic ! Shall I merge this?

**Comment by vrozic** [comment]
> LGTM, thanks @vrozic ! Shall I merge this?

Yes, it's ready to merge

**Pull request merged**